crates/ruff/src/rules/pycodestyle/logical_lines.rs

```diff
@@ -25,78 +28,327 @@ bitflags! {
     }
 }
 
+#[derive(Clone)]
+pub struct LogicalLines<'a> {
+    text: String,
+
+    /// start position, token, end position
+    tokens: Vec<(Location, &'a Tok, Location)>,
+
+    mappings: Mappings,
+
+    lines: Vec<Line>,
+}
+
+impl<'a> LogicalLines<'a> {
+    pub fn from_tokens(tokens: &'a [LexResult], locator: &Locator) -> Self {
+        assert!(u32::try_from(tokens.len()).is_ok());
+
+        let single_token = tokens.len() == 1;
+        let mut builder = LogicalLinesBuilder::with_token_capacity(tokens.len());
+        let mut parens: u32 = 0;
+
+        for (start, token, end) in tokens.iter().flatten() {
+            builder.push_token(*start, token, *end, locator);
+
+            match token {
+                Tok::Lbrace | Tok::Lpar | Tok::Lsqb => {
+                    parens += 1;
+                }
+                Tok::Rbrace | Tok::Rpar | Tok::Rsqb => {
+                    parens -= 1;
+                }
+                Tok::Newline | Tok::NonLogicalNewline | Tok::Comment(_) if parens == 0 => {
```

Comment:
for my own understanding, does this mean that the comment consumes the newline after it?

Reply:
I think so... I don't really know to be honest. I only moved this code around. 

I remember that I tried removing the comment branch because I assumed that it isn't needed because there should be a new line token after, finishing the line. But some tests started failing. It may be worth looking into this again on the top branch.